### PR TITLE
feat(locksmith) - add mark as checkin for verifiers and lock manager endpoint

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description
- add mark as checkin for verifiers and lock manager endpoint

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Review by julien51 — CHANGES_REQUESTED** [review]
Please add tests!

File: locksmith/src/controllers/v2/ticketsController.ts [hunk 1]
```diff
@@ -24,4 +24,26 @@ export class TicketsController {
     )
     response.status(200).send({ payload, signature })
   }
+
+  async markTicketAsCheckIn(request: Request, response: Response) {
+    const lockAddress = Normalizer.ethereumAddress(request.params.lockAddress)
+    const network = Number(request.params.network)
+    const metadata = request.body.message.KeyMetaData
```
Comment (julien51) [inline locksmith/src/controllers/v2/ticketsController.ts:31]: Let's not actually have `KeyMetaData` be submitted since we know the format. Let's just built it here. This way we avoid malicious behaviors.

File: locksmith/src/controllers/v2/ticketsController.ts [hunk 2]
```diff
@@ -24,13 +25,20 @@ export class TicketsController {
...
+   * This will mark a ticket as check-in, this operation is only allowed for a lock verifier of a lock manager
+   * @param {Request} request
+   * @param {Response} response
+   * @return
+   */
   async markTicketAsCheckIn(request: Request, response: Response) {
     const lockAddress = Normalizer.ethereumAddress(request.params.lockAddress)
     const network = Number(request.params.network)
-    const metadata = request.body.message.KeyMetaData
+    const payload = request.body
```
Comment (julien51) [inline locksmith/src/controllers/v2/ticketsController.ts:37]: No need for a payload anymore

File: unlock-app/src/services/storageService.ts [hunk 3]
```diff
@@ -677,4 +677,26 @@ export class StorageService extends EventEmitter {
       return false
     }
   }
+
```
Comment (julien51) [inline unlock-app/src/services/storageService.ts:680]: Let's remove this from this PR and focus _only_ on locksmith.

File: locksmith/src/utils/metadata.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+export interface KeyMetadataPayload {
+  types: {
```
[the remark this answers is not in the document]
Comment (julien51) [reply locksmith/src/utils/metadata.ts:2]: This is only useful if the user is signing things, which they are not.

File: locksmith/src/routes/v2/ticket.ts [hunk 5]
```diff
@@ -14,4 +15,12 @@ router.get(
   ticketsController.sign
 )
 
+router.put(
+  '/:network/lock/:lockAddress/key/:keyId/check',
+  isVerifierMiddleware,
```
Comment (julien51) [inline locksmith/src/routes/v2/ticket.ts:20]: Do we have this in another PR?

File: locksmith/src/controllers/v2/ticketsController.ts [hunk 6]
```diff
@@ -24,4 +25,33 @@ export class TicketsController {
...
+   * @param {Request} request
+   * @param {Response} response
+   * @return
+   */
+  async markTicketAsCheckIn(request: Request, response: Response) {
+    const lockAddress = Normalizer.ethereumAddress(request.params.lockAddress)
+    const network = Number(request.params.network)
+    const id = request.params.keyId.toLowerCase()
+
+    const metadata = generateKeyMetadataPayload(request.body).message
```
Comment (julien51) [inline locksmith/src/controllers/v2/ticketsController.ts:39]: You don't need this. You can just create the JSON object.

Comment (julien51) [reply locksmith/src/controllers/v2/ticketsController.ts:39]:
```suggestion
    const metadata = {
        checkedInAt: new Date().getTime(),
      }
```

**Review by julien51 — CHANGES_REQUESTED** [review]
Please add tests to the controller that do the following:
* test that if called without auth it fails
* test that if called from a non-verifier, it fails
* test that if called from a verifier it succeeds and updates the metadata correctly if there is existing metadata
* * test that if called from a verifier it succeeds and sets the metadata correctly if there is no metadata before

File: locksmith/src/routes/v2/ticket.ts (hunk 5 above)
Comment (kalidiagne) [reply locksmith/src/routes/v2/ticket.ts:20]: @julien51 added middleware

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 7]
```diff
@@ -65,4 +66,16 @@ describe('sign endpoint', () => {
     expect(typeof response.body.signature).toBe('string')
     expect(response.status).toBe(200)
   })
+
+  it('marks ticket as check-in', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:70]:
```suggestion
  it('marks ticket as checked-in', async () => {
```

**Comment by julien51** [comment]
@kalidiagne Please add more tests per above. It is important to avoid regression.

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 8]
```diff
@@ -67,6 +69,29 @@ describe('sign endpoint', () => {
     expect(response.status).toBe(200)
   })
 
+  it('marks ticket as checked-in without auth fails', async () => {
+    expect.assertions(1)
+    const response = await request(app).put(
+      `/v2/api/ticket/${network}/lock/${lockAddress}/key/${tokenId}/check`
+    )
+
+    expect(response.status).toBe(500)
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:78]: It should not return 500 but 401

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 9]
```diff
@@ -67,6 +69,29 @@ describe('sign endpoint', () => {
...
+    const { loginResponse } = await loginRandomUser(app)
+    expect(loginResponse.status).toBe(200)
+
+    const response = await request(app)
+      .put(
+        `/v2/api/ticket/${network}/lock/${wrongLockAddress}/key/${tokenId}/check`
+      )
+      .set('authorization', `Bearer ${loginResponse.body.accessToken}`)
+
+    expect(response.status).toBe(401)
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:92]: It should be 403

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 10]
```diff
@@ -78,4 +103,44 @@ describe('sign endpoint', () => {
...
+    expect(loginResponse.status).toBe(200)
+
+    const metadata = {}
+    const response = await request(app)
+      .put(`/v2/api/ticket/${network}/lock/${lockAddress}/key/${tokenId}/check`)
+      .set('authorization', `Bearer ${loginResponse.body.accessToken}`)
+      .send(metadata)
+    expect(response.status).toBe(202)
+
+    const updateStatus = await metadataOperations.updateKeyMetadata(metadata)
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:119]: Let's read the data rather than try to update it and check that the value is correct

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 11]
```diff
@@ -78,4 +103,44 @@ describe('sign endpoint', () => {
...
+    expect(updateStatus).toBe(false)
+  })
+
+  it('correctly marks ticket as checked-in and updates metadata', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:123]: This test should set metadata first, then check the ticket in and read the metadata to verify it is correct

File: locksmith/src/routes/v2/ticket.ts (hunk 5 above)
Comment (julien51) [inline locksmith/src/routes/v2/ticket.ts:20]: Just add the middelare for authrequired

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts (hunk 9 above)
Comment (kalidiagne) [reply locksmith/__tests__/controllers/v2/ticketsController.test.ts:92]: is not `401` correct? 

user is signed-in but, is not authorized to perform this specific request because is not a verifer

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts (hunk 8 above)
Comment (julien51) [reply locksmith/__tests__/controllers/v2/ticketsController.test.ts]: ha, we probably need to change this in another PR then.
Don't worry about it here.

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts (hunk 9 above)
Comment (kalidiagne) [reply locksmith/__tests__/controllers/v2/ticketsController.test.ts]: nice! changed also the middleware status code to match `403`

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 12]
```diff
@@ -1,12 +1,14 @@
...
+const metadataOperations = require('../../../src/operations/metadataOperations')
 
 const app = require('../../../src/app')
 
 jest.setTimeout(600000)
 const lockAddress = '0x3F09aD349a693bB62a162ff2ff3e097bD1cE9a8C'
+const wrongLockAddress = '0x00'
 const network = 4
-const tokenId = '123'
+const tokenId = '2244'
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:11]: Why the change?

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts [hunk 13]
```diff
@@ -65,4 +68,97 @@ describe('sign endpoint', () => {
...
+      .put(`/v2/api/ticket/${network}/lock/${lockAddress}/key/${tokenId}/check`)
+      .set('authorization', `Bearer ${loginResponse.body.accessToken}`)
+    expect(response.status).toBe(202)
+
+    const keyData = await metadataOperations.getKeyCentricData(
+      lockAddress,
+      tokenId
+    )
+
+    expect(keyData.metadata.value).toBe('12')
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/ticketsController.test.ts:162]: Can you also check that `custom_field` is persisted please?

File: locksmith/__tests__/controllers/v2/ticketsController.test.ts (hunk 12 above)
Comment (kalidiagne) [reply locksmith/__tests__/controllers/v2/ticketsController.test.ts:11]: now the endpoint currently avoids overrides, the old stored value for this tokenId have a not valid structure

**Review by julien51 — APPROVED** [review]
